File: images/images.py
#!/usr/bin/env python
# -*- coding: utf-8 -*

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_led_strip import LEDStrip

import sys

import config

pil_available = True
try:
    from PIL import Image
    class ImageLoaderPIL:
        image = None
        def __init__(self, f):
            self.image = Image.open(f).convert('RGB')
            size = self.get_size()
            if size != (20, 10):
                self.image = self.image.resize((20, 10), Image.ANTIALIAS)

        def get_size(self):
            if self.image == None:
                return (0, 0)

            return self.image.size

        def get_pixel(self, x, y):
            if self.image == None:
                return (0, 0, 0)

            return self.image.getpixel((x, y))
except:
    pil_available = False

    from PyQt4.QtGui import QImage, QColor
    class ImageLoaderQt:
        image = None
        def __init__(self, f):
            self.image = QImage(f)
            if self.get_size() != (20, 10):
                self.image = self.image.scaled(20, 10)

        def get_size(self):
            size = self.image.size()
            return (size.width(), size.height())

        def get_pixel(self, x, y):
            r, g, b, _ = QColor(self.image.pixel(x, y)).getRgbF()
            r, g, b = int(255*r), int(255*g), int(255*b)
            return r, g, b


if pil_available:
    ImageLoader = ImageLoaderPIL
else:
    ImageLoader = ImageLoaderQt

class Images:
    colors = [
        (10,  10,  10),  # grey
        (255, 0,   0),   # red
        (255, 80,  0),   # orange
        (255, 255, 0),   # yellow
        (0,   255, 0),   # green
        (0,   0,   255), # blue
        (255, 0,   150), # violet
        (255, 0,   40),  # purple
    ]

    leds = [x[:] for x in [[(0, 0, 0)]*config.LED_COLS]*config.LED_ROWS]
    files = []
    image_position = 0

    def __init__(self, ipcon):
        self.okay = False
        self.ipcon = ipcon

        if not config.UID_LED_STRIP_BRICKLET:
            print('Not Configured: LED Strip (required)')
            return

        self.led_strip = LEDStrip(config.UID_LED_STRIP_BRICKLET, self.ipcon)

        try:
            self.led_strip.get_frame_duration()
            print('Found: LED Strip ({0})').format(config.UID_LED_STRIP_BRICKLET)
        except:
            print('Not Found: LED Strip ({0})').format(config.UID_LED_STRIP_BRICKLET)
            return

        self.okay = True

        self.update_frame_rate()
        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         self.frame_rendered)

    def stop_rendering(self):
        if not self.okay:
            return

        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         None)

    def update_frame_rate(self):
        if not self.okay:
            return

        self.led_strip.set_frame_duration(1000.0 / config.IMAGES_FRAME_RATE)

    def new_images(self, image_urls):
        self.files = []
        for url in image_urls:
            self.files.append(ImageLoader(url))
        self.image_position = 0

    def frame_rendered(self, _):
        self.frame_upload()
        self.frame_prepare_next()

    def frame_upload(self):
        if not self.okay:
            return

        r = []
        g = []
        b = []
        for col in range(config.LED_ROWS):
            row_range = range(config.LED_COLS)
            if col % 2 == 0:
                row_range = reversed(row_range)
            for row in row_range:
                r.append(self.leds[col][row][config.R])
                g.append(self.leds[col][row][config.G])
                b.append(self.leds[col][row][config.B])

        # Make chunks of size 16
        r_chunk = [r[i:i+16] for i in range(0, len(r), 16)]
        g_chunk = [g[i:i+16] for i in range(0, len(g), 16)]
        b_chunk = [b[i:i+16] for i in range(0, len(b), 16)]

        for i in range(len(r_chunk)):
            length = len(r_chunk[i])

            # Fill up chunks with zeros
            r_chunk[i].extend([0]*(16-len(r_chunk[i])))
            g_chunk[i].extend([0]*(16-len(g_chunk[i])))
            b_chunk[i].extend([0]*(16-len(b_chunk[i])))

            try:
                self.led_strip.set_rgb_values(i*16, length, r_chunk[i], g_chunk[i], b_chunk[i])
            except:
                break

    def frame_prepare_next(self):
        if len(self.files) == 0:
            return

        h, w = self.files[self.image_position].get_size()

        self.leds = [x[:] for x in [[(0, 0, 0)]*config.LED_COLS]*config.LED_ROWS]
        for x in range(min(config.LED_COLS, w)):
            for y in range(min(config.LED_ROWS, h)):
                self.leds[y][x] = self.files[self.image_position].get_pixel(y, x)

        self.image_position = (self.image_position + 1) % len(self.files)


if __name__ == "__main__":
    ipcon = IPConnection()
    ipcon.connect(config.HOST, config.PORT)

    images = Images(ipcon)

    images.new_images(sys.argv[1:])
    images.frame_rendered(0)

    raw_input('Press enter to exit\n') # Use input() in Python 3

    ipcon.disconnect()

File: images/java/Images.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickletLEDStrip;
import com.tinkerforge.BrickletLEDStripV2;
import com.tinkerforge.TinkerforgeException;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;

class ImagesListener implements BrickletLEDStrip.FrameRenderedListener {
	private static final int CHUNK_SIZE = 16;
	private BrickletLEDStrip ledStrip = null;
	private short[][][][] images = null;
	private int imagePosition = 0;

	public ImagesListener(BrickletLEDStrip ledStrip, short[][][][] images) {
		this.ledStrip = ledStrip;
		this.images = images;
	}

	public void frameRendered(int length) {
		frameUpload();
		framePrepareNext();
	}

	private void frameUpload() {
		if (images.length == 0) {
			return;
		}

		// Reorder LED data into R, G and B channel
		short[] r = new short[Config.LED_ROWS*Config.LED_COLS];
		short[] g = new short[Config.LED_ROWS*Config.LED_COLS];
		short[] b = new short[Config.LED_ROWS*Config.LED_COLS];

		for (int row = 0, i = 0; row < Config.LED_ROWS; ++row) {
			int colBegin;
			int colEnd;
			int colStep;

			if (row % 2 == 0) {
				colBegin = Config.LED_COLS - 1;
				colEnd = -1;
				colStep = -1;
			} else {
				colBegin = 0;
				colEnd = Config.LED_COLS;
				colStep = 1;
			}

			for (int col = colBegin; col != colEnd; col += colStep) {
				r[i] = images[imagePosition][row][col][Config.R_INDEX];
				g[i] = images[imagePosition][row][col][Config.G_INDEX];
				b[i] = images[imagePosition][row][col][Config.B_INDEX];

				++i;
			}
		}

		// Make chunks of size 16
		short rChunk[] = new short[CHUNK_SIZE];
		short gChunk[] = new short[CHUNK_SIZE];
		short bChunk[] = new short[CHUNK_SIZE];

		for (int i = 0, k; i < Config.LED_ROWS*Config.LED_COLS; i += CHUNK_SIZE) {
			for (k = 0; k < CHUNK_SIZE && i + k < Config.LED_ROWS*Config.LED_COLS; ++k) {
				rChunk[k] = r[i + k];
				gChunk[k] = g[i + k];
				bChunk[k] = b[i + k];
			}

			try {
				ledStrip.setRGBValues(i, (short)k, rChunk, gChunk, bChunk);
			} catch (TinkerforgeException e) {
				break;
			}
		}
	}

	private void framePrepareNext() {
		if (images.length == 0) {
			return;
		}

		imagePosition = (imagePosition + 1) % images.length;
	}
}

class ImagesListenerV2 implements BrickletLEDStripV2.FrameStartedListener {
	private BrickletLEDStripV2 ledStripV2 = null;
	private short[][][][] images = null;
	private int imagePosition = 0;

	public ImagesListenerV2(BrickletLEDStripV2 ledStripV2, short[][][][] images) {
		this.ledStripV2 = ledStripV2;
		this.images = images;
	}

	public void frameStarted(int length) {
		frameUpload();
		framePrepareNext();
	}

	private void frameUpload() {
		int j = 0;

		if (images.length == 0) {
			return;
		}

		// Reorder LED data into R, G and B channel
		short[] r = new short[Config.LED_ROWS*Config.LED_COLS];
		short[] g = new short[Config.LED_ROWS*Config.LED_COLS];
		short[] b = new short[Config.LED_ROWS*Config.LED_COLS];
		int[] frame = new int[Config.LED_ROWS*Config.LED_COLS*3];

		for (int row = 0, i = 0; row < Config.LED_ROWS; ++row) {
			int colBegin;
			int colEnd;
			int colStep;

			if (row % 2 == 0) {
				colBegin = Config.LED_COLS - 1;
				colEnd = -1;
				colStep = -1;
			} else {
				colBegin = 0;
				colEnd = Config.LED_COLS;
				colStep = 1;
			}

			for (int col = colBegin; col != colEnd; col += colStep) {
				j = i * 3;

				r[i] = images[imagePosition][row][col][Config.R_INDEX];
				g[i] = images[imagePosition][row][col][Config.G_INDEX];
				b[i] = images[imagePosition][row][col][Config.B_INDEX];
				frame[j] = images[imagePosition][row][col][Config.R_INDEX];
				j++;
				frame[j] = images[imagePosition][row][col][Config.G_INDEX];
				j++;
				frame[j] = images[imagePosition][row][col][Config.B_INDEX];

				++i;
			}
		}

		try {
			ledStripV2.setLEDValues(0, frame);
		} catch (TinkerforgeException e) {
			return;
		}
	}

	private void framePrepareNext() {
		if (images.length == 0) {
			return;
		}

		imagePosition = (imagePosition + 1) % images.length;
	}
}

public class Images {
	private static short[][][][] images = null;
	private static IPConnection ipcon = null;
	private static BrickletLEDStrip ledStrip = null;
	private static BrickletLEDStripV2 ledStripV2 = null;
	private static ImagesListener imagesListener = null;
	private static ImagesListenerV2 imagesListenerV2 = null;

	private static BufferedImage resizeImage(BufferedImage image, int width, int height) {
		BufferedImage resizedImage = new BufferedImage(width, height, image.getType());
		Graphics2D g = resizedImage.createGraphics();

		g.drawImage(image, 0, 0, width, height, null);
		g.dispose();

		return resizedImage;
    }

	private static short[][][] readImage(String imagePath) throws Exception {
		BufferedImage image = ImageIO.read(new File(imagePath));

		if (image.getWidth() != Config.LED_ROWS ||
		    image.getHeight() != Config.LED_COLS) {
			image = resizeImage(image, Config.LED_ROWS, Config.LED_COLS);
		}

		short[][][] pixel = new short[Config.LED_ROWS][Config.LED_COLS][3];

		for (int row = 0; row < Config.LED_ROWS; ++row) {
			for (int col = 0; col < Config.LED_COLS; ++col) {
				long rgb = image.getRGB(row, col);

				pixel[row][col][0] = (short)((rgb >> 16) & 0xFF);
				pixel[row][col][1] = (short)((rgb >>  8) & 0xFF);
				pixel[row][col][2] = (short)( rgb        & 0xFF);
			}
		}

		return pixel;
	}

	public static void main(String args[]) throws Exception {
		// Create IP Connection and connect it
		ipcon = new IPConnection();
		ipcon.connect(Config.HOST, Config.PORT);

		// Call a getter to check that the Bricklet is avialable
		if (!Config.IS_LED_STRIP_V2) {
			ledStrip = new BrickletLEDStrip(Config.UID_LED_STRIP_BRICKLET, ipcon);
		}
		else {
			ledStripV2 = new BrickletLEDStripV2(Config.UID_LED_STRIP_BRICKLET, ipcon);
		}

		try {
			if (!Config.IS_LED_STRIP_V2) {
				ledStrip.getFrameDuration();
				System.out.println("Found: LED Strip " + Config.UID_LED_STRIP_BRICKLET);
			}
			else {
				ledStripV2.getFrameDuration();
				System.out.println("Found: LED Strip V2 " + Config.UID_LED_STRIP_BRICKLET);
			}
		} catch (TinkerforgeException e) {
			if (!Config.IS_LED_STRIP_V2) {
				System.out.println("Not Found: LED Strip " + Config.UID_LED_STRIP_BRICKLET);
			}
			else {
				System.out.println("Not Found: LED Strip V2 " + Config.UID_LED_STRIP_BRICKLET);
			}
			return;
		}

		// Read images
		images = new short[args.length][][][];

		for (int i = 0; i < args.length; ++i) {
			images[i] = readImage(args[i]);
		}

		// Set up listener and start rendering
		if (!Config.IS_LED_STRIP_V2) {
			ledStrip.setFrameDuration(1000 / Config.IMAGES_FRAME_RATE);
		}
		else {
			ledStripV2.setFrameDuration(1000 / Config.IMAGES_FRAME_RATE);
		}

		if (!Config.IS_LED_STRIP_V2) {
			imagesListener = new ImagesListener(ledStrip, images);
			ledStrip.addFrameRenderedListener(imagesListener);
			imagesListener.frameRendered(0);
		}
		else {
			imagesListenerV2 = new ImagesListenerV2(ledStripV2, images);
			ledStripV2.addFrameStartedListener(imagesListenerV2);
			imagesListenerV2.frameStarted(0);
		}

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}
